src/workerd/io/worker.c++

```diff
@@ -3194,44 +3192,47 @@ auto Worker::Actor::scheduleAlarm(kj::Date scheduledTime) -> kj::Promise<Schedul
   auto& scheduledAlarm = impl->maybeScheduledAlarm.emplace(
       scheduledTime, kj::newPromiseAndFulfiller<AlarmResult>());
 
-  auto whenCanceled = scheduledAlarm.resultPromise.addBranch().then(
-      [](AlarmResult result) -> ScheduleAlarmResult {
+  // Probably don't need to use kj::coCapture for this but doing so just to be on the
+  // safe side...
+  auto whenCanceled = (kj::coCapture([&scheduledAlarm]() -> kj::Promise<ScheduleAlarmResult> {
     // We've been cancelled, so return that result. Note that we cannot be resolved any other
     // way until we return an AlarmFulfiller below.
-    return result;
-  });
+    co_return co_await scheduledAlarm.resultPromise;
+  }))();
+
+  // Date.now() < scheduledTime when the alarm comes in, since we subtract elapsed CPU time from
+  // the time of last I/O in the implementation of Date.now(). This difference could be used to
+  // implement a spectre timer, so we have to wait a little longer until
+  // `Date.now() == scheduledTime`. Note that this also means that we could invoke ahead of its
+  // `scheduledTime` and we'll delay until appropriate, this may be useful in cases of clock skew.
+
+  co_return co_await handleAlarm(scheduledTime).exclusiveJoin(kj::mv(whenCanceled));
```

Comment:
@mikea it occurs to me that if the new `exclusiveJoin()` could support `ForkedPromise`s, this line could be
```cpp
co_return co_await kj::exclusiveJoin(handleAlarm(scheduleTime), scheduledAalarm.resultPromise);
```
avoiding the need for the `whenCanceled` coroutine above, or an `addBranch()`.